presto-native-execution/presto_cpp/main/operators/BroadcastExchangeSource.cpp

```diff
@@ -34,26 +34,29 @@ folly::SemiFuture<BroadcastExchangeSource::Response>
 BroadcastExchangeSource::request(
     uint32_t /*maxBytes*/,
     uint32_t /*maxWaitSeconds*/) {
-  std::vector<velox::ContinuePromise> promises;
+  if (atEnd_) {
```

Comment:
We access atEnd_, reader_ w/o a lock?
Is this ok?
What is std::lock_guard<std::mutex> l(queue_->mutex()); suppose to protect?

Reply:
It is supposed to protect access to the `queue` which can happen from multiple threads (e.g.: when different ExchangeSource's trying to add some data into the queue).

The `atEnd` field is a local field of an `ExchangeSource`. In my understanding `ExchangeSource` should not be accessed concurrently from multiple threads. The `ExchangeClient` picks an `ExchangeSource` to load under a lock and removes it from the queue to prevent other threads from calling into the same `ExchangeSource`.